src/main/daml/Daml/Finance/Account/Account.daml

```diff
@@ -23,38 +23,62 @@ template Account
     owner : Party
       -- ^ The account owner.
     controllers : Account.Controllers
-      -- ^ Controllers of transfers.
+      -- ^ Controllers of transfers and freezes.
     id : Id
       -- ^ Identifier of the account.
     description : Text
       -- ^ Description of the account.
     holdingFactoryCid : ContractId HoldingFactory.F
       -- ^ Associated holding factory.
+    unfreezers : Parties
+      -- ^ Parties who has frozen the account and are authorized to unfreeze it.
     observers : PartiesMap
       -- ^ Observers.
   where
-    signatory custodian, owner
+    signatory custodian, owner, unfreezers
     observer Disclosure.flattenObservers observers
 
     -- Outgoing controllers must be non-empty.
     ensure not . S.null $ controllers.outgoing
 
-    let account = AccountKey with custodian; owner; id
+    let
+      account = AccountKey with custodian; owner; id
+      mustNotBeFrozen = assertMsg @Update "account must not be frozen" $ S.null unfreezers
+      contextFrozen = "Frozen"
 
     interface instance Account.I for Account where
-      view = Account.View with custodian; id; owner; holdingFactoryCid; description; controllers
+      view = Account.View with
+        custodian; id; owner; holdingFactoryCid; description; controllers; unfreezers
       getKey = account
-      credit Account.Credit{quantity} =
+      credit Account.Credit{quantity} = do
+        mustNotBeFrozen
         exercise holdingFactoryCid HoldingFactory.Create with
           instrument = quantity.unit
           account
           amount = quantity.amount
           observers = M.empty
       debit Account.Debit{holdingCid} = do
+        mustNotBeFrozen
         vHolding <- view <$> fetch holdingCid
         assertMsg "accounts must match" $ vHolding.account == account
         exercise holdingFactoryCid HoldingFactory.Remove with
           actors = S.fromList [account.custodian, account.owner]; holdingCid
+      freeze Account.Freeze{freezers} = do
+        mustNotBeFrozen
+        assertMsg "Freezers must be non-empty" (not . S.null $ freezers)
+        newCid <- toInterfaceContractId @Disclosure.I <$> create this with unfreezers = freezers
+        -- adding freezers as observers to Account.R
+        coerceInterfaceContractId @Account.I <$> exercise newCid Disclosure.AddObservers with
+          disclosers = S.fromList (signatory this) <> freezers
+          observersToAdd = (contextFrozen, freezers)
+      unfreeze = do
+        newCid <- toInterfaceContractId <$> create this with unfreezers = S.empty
+        -- removing freezers as observers to Account.R
+        optional newCid coerceContractId <$>
+          exercise (toInterfaceContractId @Disclosure.I newCid)
+            Disclosure.RemoveObservers with
+              disclosers = S.fromList (signatory this) <> unfreezers
+              observersToRemove = (contextFrozen, unfreezers)
```

Comment:
You should first check if the account is actually frozen before preforming this action.

Reply:
Actually it is not needed, as the choice can only be exercised if `unfreezers` are non-empty, i.e., when it is frozen. But agree that it is easier to understand the code with such a check, and we could then also add a better error message.